porch/controllers/rootsyncsets/pkg/controllers/rootsyncset/controller.go

```diff
@@ -108,59 +116,215 @@ func (r *RootSyncSetReconciler) Reconcile(ctx context.Context, req ctrl.Request)
+	}
+	data, err := json.Marshal(newRootSync)
+	if err != nil {
+		return fmt.Errorf("failed to encode root sync to JSON: %w", err)
+	}
+	rs, err := client.Resource(rootSyncGVR).Namespace(rootSyncNamespace).Patch(ctx, name, types.ApplyPatchType, data, metav1.PatchOptions{FieldManager: name})
+	if err != nil {
+		return fmt.Errorf("failed to patch RootSync %w", err)
+	}
+	klog.Infof("Create/Update resource %s as %v", name, rs)
+	return nil
+}
+
+// checkSyncStatus fetches the RootSync using the provided client and computes the sync status. The rules
+// for computing status here mirrors the one used in the status command in the nomos cli.
+func (r *RootSyncSetReconciler) checkSyncStatus(ctx context.Context, client dynamic.Interface, rssName string) (string, error) {
+	rs, err := client.Resource(rootSyncGVR).Namespace(rootSyncNamespace).Get(ctx, rssName, metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("failed to get RootSync %w", err)
+	}
+
+	conditions, _, err := unstructured.NestedSlice(rs.Object, "status", "conditions")
+	if err != nil {
+		return "", fmt.Errorf("failed to extract conditions from RootSync: %w", err)
+	}
+
+	val, found, err := getConditionStatus(conditions, "Stalled")
+	if err != nil {
+		return "", fmt.Errorf("error fetching condition 'Stalled' from conditions slice: %w", err)
+	}
+	if found && val == "True" {
+		return "Stalled", nil
+	}
+
+	val, found, err = getConditionStatus(conditions, "Reconciling")
+	if err != nil {
+		return "", fmt.Errorf("error fetching condition 'Reconciling' from conditions slice: %w", err)
+	}
+	if found && val == "True" {
+		return "Reconciling", nil
+	}
+
+	cond, found, err := getCondition(conditions, "Syncing")
+	if err != nil {
+		return "", fmt.Errorf("error fetching condition 'Syncing' from conditions slice: %w", err)
+	}
+	if !found {
+		return "Reconciling", nil
+	}
+
+	errCount, err := extractErrorCount(cond)
+	if err != nil {
+		return "", fmt.Errorf("error extracting error count from 'Syncing' condition: %w", err)
+	}
+	if errCount > 0 {
+		return "Error", nil
+	}
+
+	val, err = extractStringField(cond, "status")
+	if err != nil {
+		return "", fmt.Errorf("error extracting status of 'Syncing' condition: %w", err)
+	}
+	if val == "True" {
+		return "Pending", nil
+	}
+
+	return "Synced", nil
+}
+
+func getConditionStatus(conditions []interface{}, condType string) (string, bool, error) {
+	cond, found, err := getCondition(conditions, condType)
+	if err != nil {
+		return "", false, err
+	}
+	if !found {
+		return "", false, nil
+	}
+	s, err := extractStringField(cond, "status")
 	if err != nil {
-		return schema.GroupVersionResource{}, nil, fmt.Errorf("failed to parse group version when building object: %w", err)
+		return "", false, err
 	}
-	rootSyncRes := schema.GroupVersionResource{Group: gv.Group, Version: gv.Version, Resource: "rootsyncs"}
+	return s, true, nil
+}
+
+func getCondition(conditions []interface{}, condType string) (map[string]interface{}, bool, error) {
```

Comment:
Do we have the type (or could we embed at least the bit we care about?)  If so, it might then be easier to use `runtime.DefaultUnstructuredConverter.FromUnstructured`

Reply:
Yeah, now that Config Sync is OSS we can use the types directly here.